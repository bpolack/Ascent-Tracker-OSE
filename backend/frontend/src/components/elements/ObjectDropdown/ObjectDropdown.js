import React, { Component } from 'react';
import { Menu, Transition } from '@headlessui/react';

// Fontawesome
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronDown, faTrash, faPencilAlt } from '@fortawesome/free-solid-svg-icons';

export class ObjectDropdown extends Component {
	render() {
		return (
			<Menu as="div" className="absolute right-3 top-3 inline-block text-left">
				{({ open }) => (
					<>
						<Menu.Button className="inline-flex justify-center items-center w-full px-4 py-2 text-sm font-medium text-white dark:text-gray-800 bg-gray-800 dark:bg-white rounded-full bg-opacity-60 hover:bg-opacity-90 focus:outline-none focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75">
							Options
							<FontAwesomeIcon icon={faChevronDown} className="ml-2" />
						</Menu.Button>
						<Transition
							show={open} 
							enter="transition ease-out duration-100"
							enterFrom="transform opacity-0 scale-95"
							enterTo="transform opacity-100 scale-100"
							leave="transition ease-in duration-75"
							leaveFrom="transform opacity-100 scale-100"
							leaveTo="transform opacity-0 scale-95"
						>
							<Menu.Items static className="absolute right-0 w-56 mt-2 origin-top-right bg-white divide-y divide-gray-100 rounded-md shadow-lg ring-1 ring-black ring-opacity-10 focus:outline-none">
								<div className="px-1 py-1 ">
									<Menu.Item onClick={this.props.handleEdit}>
										{({ active }) => (
											<button
												className={`${active ? 'bg-pink-500 text-white' : 'text-gray-900'
													} group flex rounded-md items-center w-full px-2 py-2 text-sm`}
											>
												<FontAwesomeIcon icon={faPencilAlt} className="mr-3" />
												Edit
											</button>
										)}
									</Menu.Item>
								</div>
								<div className="px-1 py-1">
									<Menu.Item onClick={this.props.handleDelete}>
										{({ active }) => (
											<button
												className={`${active ? 'bg-pink-500 text-white' : 'text-gray-900'
													} group flex rounded-md items-center w-full px-2 py-2 text-sm`}
											>
												<FontAwesomeIcon icon={faTrash} className="mr-3" />
												Delete
											</button>
										)}
									</Menu.Item>
								</div>
							</Menu.Items>
						</Transition>
					</>
      			)}
			</Menu>
		)
	}
}

export default ObjectDropdown
