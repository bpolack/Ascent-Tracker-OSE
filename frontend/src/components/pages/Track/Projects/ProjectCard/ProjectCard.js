import React, { Component } from 'react'

export class ProjectCard extends Component {
	render() {
		return (
			<div>
				
			</div>
		)
	}
}

export default ProjectCard
