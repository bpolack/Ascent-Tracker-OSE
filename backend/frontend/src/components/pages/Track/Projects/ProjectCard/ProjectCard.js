import React, { Component } from 'react'

// Import Required Components
import ObjectDropdown from '../../../../elements/ObjectDropdown/ObjectDropdown';

export class ProjectCard extends Component {
	constructor(props) {
		super(props);

		this.handleEdit = this.handleEdit.bind(this);
		this.handleDelete = this.handleDelete.bind(this);
	}

	handleEdit() {
		const { _id } = this.props.project;
		this.props.openEditModal(_id);
	}

	handleDelete() {
		const { _id } = this.props.project;
		this.props.openDeleteModal(_id);
	}

	render() {

		const { name, desc } = this.props.project;

		return (
			<div className="relative bg-ascent-std text-ascent-std shadow transition-all p-6 rounded-xl">
				<h2 className="heading-ascent-3 mt-8">{name}</h2>
				<p className="mb-2">{desc}</p>
				<ObjectDropdown handleEdit={this.handleEdit} handleDelete={this.handleDelete} />
			</div>
		)
	}
}

export default ProjectCard
