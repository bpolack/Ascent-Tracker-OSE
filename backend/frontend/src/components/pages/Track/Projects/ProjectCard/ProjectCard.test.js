import React from 'react';
import { shallow } from 'enzyme';
import { ProjectCard } from './ProjectCard';
import ObjectDropdown from '../../../../elements/ObjectDropdown/ObjectDropdown';

describe('ProjectCard', () => {
	it("Renders correctly without crashing", () => {
		const openEditMock = jest.fn();
		const openDeleteMock = jest.fn();
		const project = {
			_id: "asdf1234",
			name: "Cart Title",
			desc: "Testing this card"
		};

		const wrapper = shallow(<ProjectCard openEditModal={openEditMock} openDeleteModal={openDeleteMock} project={project} />);
		expect(wrapper).toMatchSnapshot();
	});
	describe('When user clicks edit', () => {
		it("invokes the passed openModal prop for editing", () => {
			const openEditMock = jest.fn();
			const openDeleteMock = jest.fn();
			const project = {
				_id: "asdf1234",
				name: "Cart Title",
				desc: "Testing this card"
			};

			const wrapper = shallow(<ProjectCard openEditModal={openEditMock} openDeleteModal={openDeleteMock} project={project} />);

			const dropdownElement = wrapper.find(ObjectDropdown);
			dropdownElement.invoke('handleEdit')();

			expect(openEditMock).toHaveBeenCalledTimes(1);
			expect(openEditMock).toHaveBeenCalledWith("asdf1234");
		});
	});

	describe('When user clicks delete', () => {
		it("invokes the passed openModal prop for deleting", () => {
			const openEditMock = jest.fn();
			const openDeleteMock = jest.fn();
			const project = {
				_id: "asdf1234",
				name: "Cart Title",
				desc: "Testing this card"
			};

			const wrapper = shallow(<ProjectCard openEditModal={openEditMock} openDeleteModal={openDeleteMock} project={project} />);

			const dropdownElement = wrapper.find(ObjectDropdown);
			dropdownElement.invoke('handleDelete')();

			expect(openDeleteMock).toHaveBeenCalledTimes(1);
			expect(openDeleteMock).toHaveBeenCalledWith("asdf1234");
		});
	});
});