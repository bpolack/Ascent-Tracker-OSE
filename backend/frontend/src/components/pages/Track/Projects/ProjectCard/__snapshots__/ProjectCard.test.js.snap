// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard Renders correctly without crashing 1`] = `ShallowWrapper {}`;
